### PR TITLE
[RLlib] Fix torch scheduler stepping and reporting.

### PR DESCRIPTION
**Pull request by simonsays1980**
## Why are these changes needed?

When using `torch.optim.lr_scheduler.LRScheduler` schedulers learning rates were wrongly updated and reported b/c they were stepped in `TorchLearner.apply_gradients` and reported in `Learner.after_gradient_based_update`. The latter function also updated any RLlib specific learning rate schedules and therefore __reset__ any learning rate schedules of the torch-specific schedulers. 

This PR fixes this error and instead overrides the `after_gradient_based_update` to step learning rate schedulers in there and report them correctly. It also avoids the RLlib schedulers to override updated learning rates. In case no torch-specific learning rates are used, the `super()` is called an RLlib schedules do their work. 

This setup also updates the torch-specific schedulers at the right point in regard to SGD: after an SGD epoch and not in each minibatch step.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [x] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/core/learner/torch/torch_learner.py [hunk 1]
```diff
@@ -257,15 +262,57 @@ def apply_gradients(self, gradients_dict: ParamDict) -> None:
...
                     )
 
-                    # If the module uses learning rate schedulers, step them here.
-                    if module_id in self._lr_schedulers:
-                        for scheduler in self._lr_schedulers[module_id][optimizer_name]:
-                            scheduler.step()
-
-                    # If the module uses learning rate schedulers, step them here.
-                    if module_id in self._lr_schedulers:
+    @OverrideToImplementCustomLogic_CallToSuperRecommended
```
Comment (sven1977) [inline rllib/core/learner/torch/torch_learner.py:265]: Great! This is so much better to do this in this method here (instead of in `apply_gradients`).

File: rllib/core/learner/torch/torch_learner.py [hunk 2]
```diff
@@ -257,15 +262,57 @@ def apply_gradients(self, gradients_dict: ParamDict) -> None:
...
+                        key=(
+                            module_id,
+                            f"{optimizer_name[len(module_id) + 1:]}_{LR_KEY}",
+                        ),
+                        value=convert_to_numpy(self._get_optimizer_lr(optimizer)),
+                        window=1,
+                    )
+        # Otherwise call the `super()`'s method to update RLlib's learning rate
+        # schedules.
+        else:
```
Comment (sven1977) [inline rllib/core/learner/torch/torch_learner.py:314]: super nit: Can we flip this if-block and move the default RLlib case to the top? Then we don't even need an else block anymore and can save the long indentation for the rest of this method.

File: rllib/examples/learners/ppo_with_torch_lr_schedulers.py [hunk 3]
```diff
@@ -11,8 +11,8 @@
 
 How to run this script
 ----------------------
-`python [script file name].py --enable-new-api-stack --lr-const-factor=0.1
---lr-const-iters=10 --lr-exp-decay=0.3`
+`python [script file name].py --enable-new-api-stack --lr-const-factor=0.9
```
Comment (sven1977) [inline rllib/examples/learners/ppo_with_torch_lr_schedulers.py:14]: Could we add a quick custom callback here that checks the learning rate after each iteration?

Roughly like this:
```
class LRChecker(DefaultCallbacks):
   def on_train_result(self, result, algorithm, ...):
      assert algorithm.learner_group._learner.get_optimizer(DEFAULT_MODULE_ID)... param_groups.... ["lr"] == [some value based on `algorithm.iteration`]
```

It's ok to do this check only for the first 2 or 3 iterations.

**Review by sven1977 — COMMENTED** [review]
LGTM. Awesome fix, thanks @simonsays1980 . Just one important request about adding the assertion into the example script such that this script really acts as a unit test for this functionality.

File: rllib/core/learner/torch/torch_learner.py (hunk 2 above)
Comment (simonsays1980) [reply rllib/core/learner/torch/torch_learner.py:314]: Will do. Makes totally sense.

File: rllib/examples/learners/ppo_with_torch_lr_schedulers.py [hunk 4]
```diff
@@ -49,26 +49,88 @@
...
 )
+from ray.rllib.utils.metrics.metrics_logger import MetricsLogger
 from ray.rllib.utils.framework import try_import_torch
 from ray.rllib.utils.test_utils import add_rllib_example_script_args
 
 torch, _ = try_import_torch()
 
-parser = add_rllib_example_script_args(default_reward=450.0, default_timesteps=200000)
+
+class LRChecker(DefaultCallbacks):
```
Comment (sven1977) [inline rllib/examples/learners/ppo_with_torch_lr_schedulers.py:74]: Nice!

Comment (sven1977) [inline rllib/examples/learners/ppo_with_torch_lr_schedulers.py:74]:
```suggestion
class LRChecker(DefaultCallbacks):
    """A custom callback that asserts the functionality of the lr-scheduler used."""
```

File: rllib/core/learner/torch/torch_learner.py [hunk 5]
```diff
@@ -271,15 +276,56 @@ def apply_gradients(self, gradients_dict: ParamDict) -> None:
...
-                    if module_id in self._lr_schedulers:
-                        for scheduler in self._lr_schedulers[module_id][optimizer_name]:
-                            scheduler.step()
+        # Only update this optimizer's lr, if a scheduler has been registered
+        # along with it.
+        for module_id, optimizer_names in self._module_optimizers.items():
+            for optimizer_name in optimizer_names:
+                # If learning rate schedulers are provided step them here. Note,
+                # stepping them in `TorchLearner.apply_gradients` updates the
+                # learning rates during minibatch updates; we want to update
```
Comment (sven1977) [inline rllib/core/learner/torch/torch_learner.py:310]: thanks for clarifying this.

**Review by sven1977 — APPROVED** [review]
LGTM now! Thanks for the additional fixes and adding the test case (through the callback) to the example script! Very cool.
